### PR TITLE
Fix typing & add some tests

### PR DESCRIPTION
**Pull request by zundertj**
A number of `# type: ignore` statements were not needed any more, or could easily be fixed. Also added some tests to cover some corner cases not tested yet.

### CODE REVIEW

File: py-polars/polars/cfg.py [hunk 1]
```diff
@@ -14,7 +14,7 @@ def set_utf8_tables(cls) -> "Type[Config]":
         """
         Use utf8 characters to print tables
         """
-        os.environ.unsetenv("POLARS_FMT_NO_UTF8")  # type: ignore
+        os.environ["POLARS_FMT_NO_UTF8"] = ""
```
Comment (ritchie46) [inline py-polars/polars/cfg.py:17]: I don't think this is the same. We check if the `env var` exists from the rust side. Now it does exist, where we first removed it.

File: py-polars/polars/internals/construction.py [hunk 2]
```diff
@@ -110,14 +110,9 @@ def sequence_to_pyseries(
...
-        if dtype == Date:
-            pyseries = pyseries.cast(str(Date), True)
-        elif dtype == Datetime:
-            pyseries = pyseries.cast(str(Datetime), True)
-        elif dtype == Time:
-            pyseries = pyseries.cast(str(Time), True)
-        elif dtype == Categorical:
-            pyseries = pyseries.cast(str(Categorical), True)
+
+        if dtype in (Date, Datetime, Time, Categorical):
```
Comment (ritchie46) [inline py-polars/polars/internals/construction.py:114]: Way better!

File: py-polars/polars/internals/expr.py [hunk 3]
```diff
@@ -796,7 +803,7 @@ def shift(self, periods: int = 1) -> "Expr":
         """
         return wrap_expr(self._pyexpr.shift(periods))
 
-    def shift_and_fill(self, periods: int, fill_value: "Expr") -> "Expr":
+    def shift_and_fill(self, periods: int, fill_value: Union[int, "Expr"]) -> "Expr":
```
Comment (ritchie46) [inline py-polars/polars/internals/expr.py:806]: I think that it may be any value that we can turn into an `Expr` literal, so `flt, bool, int, str`

Comment (zundertj) [reply py-polars/polars/internals/expr.py:806]: I was thinking the same, did here the minimum to make `mypy` happy. Should it be `Any` in that case, or do we want to limit to a subset (bool, int, float, str, date, ?)?

File: py-polars/polars/cfg.py (hunk 1 above)
Comment (zundertj) [reply py-polars/polars/cfg.py:17]: I should have added a comment here when making this change, as it is not obvious. 

There are a couple of things going on here:
1. `unsetenv` lives in the documentation on `os`, although it seems it also lives on `os.environ`. I think this is also why `mypy` otherwise complains, it doesn't validate.
2. on Windows, it is only implemented from Python 3.9 onwards, see https://docs.python.org/3.10/library/os.html#os.unsetenv
3. the recommendation in the docs say it is probably best to do `del os.environ["POLARS_FMT_NO_UTF8"]`
4. in the bug tracker, it is pointed out that it is implemented in Python 3.8 as I did here, see top comment on https://bugs.python.org/issue39413

Overall, I would say the better fix is to not rely on removing an ENV variable, but explicitly set 0 or 1. I.e., below there is another config that says this:

```python
 @classmethod
    def set_utf8_tables(cls) -> "Type[Config]":
        """
        Use utf8 characters to print tables
        """
        os.environ.unsetenv("POLARS_FMT_NO_UTF8")  # type: ignore
        return cls

    @classmethod
    def set_ascii_tables(cls) -> "Type[Config]":
        """
        Use ascii characters to print tables
        """
        os.environ["POLARS_FMT_NO_UTF8"] = "1"
        return cls
```

it would be more consistent if we could do `os.environ["POLARS_FMT_NO_UTF8"] = "0" in what I changed above, but Im not sure how things are handled on the Rust side.

File: py-polars/polars/internals/expr.py [hunk 4]
```diff
@@ -1675,7 +1661,9 @@ def rank(self, method: str = "average") -> "Expr":  # type: ignore
         """
         return wrap_expr(self._pyexpr.rank(method))
 
-    def diff(self, n: int = 1, null_behavior: str = "ignore") -> "Expr":  # type: ignore
+    def diff(
+        self, n: int = 1, null_behavior: Literal["ignore", "drop"] = "ignore"
```
Comment (ghuls) [inline py-polars/polars/internals/expr.py:1665]: `null_behavior` should alosohave `str` as type as with `Literal` only .

```python
my_variable = "ignore"

diff (..., null_behavior=my_variable)
```
As using a variable with that string inside, will not typecheck.

Not sure if we better use an Union or an overload
https://www.python.org/dev/peps/pep-0586/#interactions-with-overloads

Comment (zundertj) [reply py-polars/polars/internals/expr.py:1665]: That is an important gotcha. I was under the impression that would pass, and that therefore `Literal` would help us to tighten up the types by changing all string options into Literals...

File: py-polars/polars/internals/expr.py [hunk 5]
```diff
@@ -1244,27 +1251,6 @@ def is_between(
             expr = self
         return ((expr > start) & (expr < end)).alias("is_between")
 
-    @property
```
Comment (zundertj) [inline py-polars/polars/internals/expr.py:1247]: Moved down, see comment in code where it is inserted.

File: py-polars/polars/internals/expr.py (hunk 3 above)
Comment (ritchie46) [reply py-polars/polars/internals/expr.py:800]: I think limit to that subset, for other types we cannot create a `pl.lit`

File: py-polars/polars/cfg.py (hunk 1 above)
Comment (ghuls) [reply py-polars/polars/cfg.py:17]: At the moment it just checks if this environment variable exists.
```rust
./polars/polars-core/src/fmt.rs
369-        #[cfg(feature = "pretty_fmt")]
370-        {
371-            let mut table = Table::new();
372:            let preset = if std::env::var("POLARS_FMT_NO_UTF8").is_ok() {
373-                ASCII_FULL
374-            } else {
375-                UTF8_FULL
```

Comment (zundertj) [reply py-polars/polars/cfg.py]: I have reverted this, the initial goal of removing the `type: ignore` becomes muddied with the inconsistent handling between the various configs; I feel that is better tackled in a separate PR.

**Comment by ritchie46** [comment]
Let me know when its good to go.

**Comment by zundertj** [comment]
> Let me know when its good to go.

All good now, just pushed the update to allow other types.
